README.md
[![Please do not theme this app](https://stopthemingmy.app/badge.svg)](https://stopthemingmy.app)

# Apostrophe

![](screenshots/main.png)

## About

Apostrophe is a [GTK+](https://www.gtk.org) based distraction free Markdown editor, originally developed by Wolf Vollprecht and currently developed and maintained by Manuel Genovés. It uses pandoc as back-end for parsing Markdown and offers a very clean and sleek user interface.

## Install

<a href='https://flathub.org/apps/details/org.gnome.gitlab.somas.Apostrophe'><img width='240' alt='Download on Flathub' src='https://flathub.org/assets/badges/flathub-badge-en.png'/></a>

Also several unofficial builds are available:

* Nix(OS): [`pkgs.apostrophe`](https://github.com/NixOS/nixpkgs/blob/master/pkgs/applications/editors/apostrophe/default.nix)
* Arch Linux (AUR)
* [Fedora](https://src.fedoraproject.org/rpms/apostrophe): `sudo dnf install apostrophe`

## Translation

If you want to help translating the project, please join us at [Damned Lies](https://l10n.gnome.org/module/apostrophe/).
Any help is appreciated!

## Building

### Building using GNOME Builder

GNOME Builder offers the easiest method to build Apostrophe. Just follow [this guide](https://wiki.gnome.org/Newcomers/BuildProject) and you'll be up and running in a minute.

### Building from Git

To build Apostrophe from source you need to have the following dependencies installed:

- Build system: `meson ninja-build`
- Pandoc, the program used to convert Markdown to basically anything else: `pandoc`
- GTK3 and GLib development packages: `libgtk-3-dev libglib2.0-dev`
- Rendering the preview panel: `libwebkit2gtk`
- Spell checking libraries: `gobject-introspection libgirepository1.0-dev gir1.2-gspell1 gettext`
- Python dependencies: `python3 python3-regex python3-setuptools python3-levenshtein python3-enchant python3-gi python3-cairo python3-pypandoc`
- *optional:* AppStream utility: `appstream-util`
- *optional:* pdftex module: `texlive texlive-latex-extra`
- *optional:* formula preview: `libjs-mathjax`

Depending on your setup you may need to install these schemas before building:

```bash
$ sudo cp data/org.gnome.gitlab.somas.Apostrophe.gschema.xml /usr/share/glib-2.0/schemas/org.gnome.gitlab.somas.Apostrophe.gschema.xml
$ sudo glib-compile-schemas /usr/share/glib-2.0/schemas
```

Once all dependencies are installed you can build Apostrophe using the following commands:

```bash
$ git clone https://gitlab.gnome.org/World/apostrophe/
$ cd apostrophe
$ meson builddir --prefix=/usr -Dprofile=development
$ sudo ninja -C builddir install
```

Then you can run the installed package:

```bash
$ apostrophe
```

Or a local version which runs from the source tree
```bash
$ ./builddir/local-apostrophe
```



### Building a flatpak package

It's also possible to build, run and debug a flatpak package. All you need is to setup [flatpak-builder](https://docs.flatpak.org/en/latest/first-build.html) and run the following commands:

```bash
$ cd build-aux/flatpak
$ flatpak-builder --force-clean --install --user _build org.gnome.gitlab.somas.Apostrophe.json
```


NEWS
Version 3.0.alpha
~~~~~~~~~~~~~~~
Released:

Features:
- GTK4 port
- Live stats for selected text
- Improved UI and styling
- Improved Hemingway Mode
- Implemented notification on external file changes

Version 2.6.3
~~~~~~~~~~~~~~~
Released: 2022-04-29

Bugfixes:
- Fix a bug where the preview wouldn't be shown when opening a file from Files or the CLI
- Fix a bug where horizontal preview wouldn't request enough space at startup
- Removed swipe gestures from the preview panel

Features:
- Added Persian translations, updated Catalan, German , Finnish and Turkish ones

Version 2.6.2
~~~~~~~~~~~~~~~
Released: 2022-03-31

Bugfixes:
- Fix handling of windowed preview mode on some circumstances

Features:
- Add proper titles to the preview window
- Update translations

Version 2.6.1
~~~~~~~~~~~~~~~
Released: 2022-03-20

Bugfixes:
- A small bug in 2.6 which didn't allow to save files

Version 2.6
~~~~~~~~~~~~~~~
Released: 2022-03-19

Features:
- Implemented multiwindow support
- Save window state on exit
- Ported to libhandy
- Improved the overall UI
- New animations
- Follow the Dark Mode global preference
- Add an "automatic" color mode
- Improve the High Contrast styles
- Improve inline preview markup

Version 2.5
~~~~~~~~~~~~~~~
Released: 2021-08-28

Features:
- Added Sepia Mode
- Improved exported HTML on small screens
- Updated metadata and runtime version

Version 2.4
~~~~~~~~~~~~~~~
Released: 2021-03-10

Bugfixes:
- Fixed a bug where the cursor could go over the bottom of the window
- Fixed a bug where in certain circumstances the bottom line couldn't be clicked
- Improved behaviour when dealing with unsaved files.
- Improved start/end sentence detection in Focus Mode
- Small improvements in the headerbar title
- Cleaned a little bit the Shortcuts Window

Version 2.3
~~~~~~~~~~~~~~~
Released: 2021-03-01

Features:
- New export dialog
- Improved Reveal.js slideshow exports
- Better input/output handling
- Inhibit logout if the document is not saved
- Migrate to xelatex
- Better internal path handling
- New animated progressbar on saving
- Strict sandboxing support

Bugfixes:
- Undo/redo improvements
- Fixes for bad markdown parsing of unordered lists

Version 2.2.0.3
~~~~~~~~~~~~~~~
Released: 2020-05-04

Bugfixes:
 - Better wording for some areas
 - Fixed some issues where Apostrophe couldn't open files from CLI/context menu
 - Now Apostrophe checks if there is unsaved work before exiting via CTRL + Q/W
 - Better parsing of weblinks and links to files
 - Fixed a bug where the bottom portion of the screen was insensitive to mouse input

Version 2.2.0.2
~~~~~~~~~~~~~~~
Released: 2020-04-20

Bugfixes:
 - Fixes for the Elementary OS platform
 - Fixed some links and metadata

Version 2.2.0.1
~~~~~~~~~~~~~~~
Released: 2020-04-18

Features:
 - New headerbar design
 - New preview modes, with the option to sync them to the edit view
 - New preview mode selector
 - New theme selector
 - Rework the autohiding mechanism; now the headerbar fades away when typing, only to reappear when the cursor moves to the top portion of the window
 - Now the content of the texview goes visually bellow the headerbar
 - Overall better styling
 - Added Hemingway mode, which disables the backspace key
 - Added Github Flavoured Markdow, MultiMarkdown, Pandoc's Markdown and Commonmark support, being CommonMark the default from now on
 - New stats counter, with the option to show count of characters/words/sentences/paragrafs/reading time
 - Better handling of DnD events
 - Export to A4 by default
 - Technical improvements 
    * Port of the buildsystem to Meson. Now you can hit the "build" button on Builder and everything works as expected
    * Port to gspell
    * Partial port to gresources
    * Overall refactorization of the codebase
    * General bugfixes and improvements

Version 2.1.5
~~~~~~~~~~~~~
Released: 2019-03-10

Features:
 - Added italian language
 - Initial themes support: now apostrophe adapts his colors to the current GTK theme
 - Disabled scroll gradient, can be enabled in the preferences dialog
 - Allow to disable headerbar autohidding in Dconf
 - Now a single click is enough to open files in the recent files popover
 - Spellchecking status is now saved between sessions
 - Minor UI fixes
 - Added -d flag to enable webdev tools

Version 2.1.4
~~~~~~~~~~~~~
Released: 2018-12-06

Features:
 - Updated css styles.

Version 2.1.3
~~~~~~~~~~~~~
Released: 2018-11-28

Features:
 - This release features a new logo, polishes the Appmenu, fixes usability bugs and flatpak related bugs.

Version 2.1.2
~~~~~~~~~~~~~
Released: 2018-07-27

Features:
 - This release provides a fix to a bug that caused Apostrophe to not mark properly **bold**, *cursive*, and ***bold and cursive*** words.

Version 2.1.1
~~~~~~~~~~~~~
Released: 2018-07-26

Bugfixes:
 - One on focus mode which caused the lines to be highlighted on edit rather than on click
 - Non symbolic icons on the searchbar

Version 2.1.0
~~~~~~~~~~~~~
Released: 2018-07-18

Features:
 - Drop AppMenu support
 - HeaderBar and menus redesign, with a new unified menu and quick access buttons on the headerbar
 - Now the fullscreen view shows a headerbar when the cursor approaches the top of the screen
 - A new unified export dialog, with updated options, and quick access to pdf, odt and html export
 - Bugfixes.

Version 2.0.4
~~~~~~~~~~~~~
Released: 2018-06-24

Features:
 - Now the menu is a Popover instead a regular menu.
 - The headerbar matches the theme selected for the application.
 - Updated translations.

Version 2.0.3
~~~~~~~~~~~~~
Released: 2018-06-14

Bugfixes:
 - Small bug fixes, updated links.

Version 2.0.2
~~~~~~~~~~~~~
Released: 2018-05-16

Bugfixes:
 - Fix a bug with the preview mode.

Version 2.0.1
~~~~~~~~~~~~~
Released: 2018-05-14

Bugfixes:
 - Don’t use env variable to check if in flatpak.

Version 2.0.0
~~~~~~~~~~~~~
Released: 2018-05-12

Features:
 - First re-release


assets/js/main.js
/*
	Story by HTML5 UP
	html5up.net | @ajlkn
	Free for personal and commercial use under the CCA 3.0 license (html5up.net/license)
*/

(function($) {

	var	$window = $(window),
		$body = $('body'),
		$wrapper = $('#wrapper');

	// Breakpoints.
		breakpoints({
			xlarge:   [ '1281px',  '1680px' ],
			large:    [ '981px',   '1280px' ],
			medium:   [ '737px',   '980px'  ],
			small:    [ '481px',   '736px'  ],
			xsmall:   [ '361px',   '480px'  ],
			xxsmall:  [ null,      '360px'  ]
		});

		// hide .navbar first
		$(".navbar").hide();

		// fade in .navbar
		$(function () {
			$(window).scroll(function () {
	
		// set distance user needs to scroll before we start fadeIn
				if ($(this).scrollTop() > $window.height()) {
					$('.navbar').fadeIn();
				} else {
					$('.navbar').fadeOut();
				}
			});
		});

	// Play initial animations on page load.
		$window.on('load', function() {
			window.setTimeout(function() {
				$body.removeClass('is-preload');
			}, 100);
		});

	// Browser fixes.

		// IE: Flexbox min-height bug.
			if (browser.name == 'ie')
				(function() {

					var flexboxFixTimeoutId;

					$window.on('resize.flexbox-fix', function() {

						var $x = $('.fullscreen');

						clearTimeout(flexboxFixTimeoutId);

						flexboxFixTimeoutId = setTimeout(function() {

							if ($x.prop('scrollHeight') > $window.height())
								$x.css('height', 'auto');
							else
								$x.css('height', '100vh');

						}, 250);

					}).triggerHandler('resize.flexbox-fix');

				})();

		// Object fit workaround.
			if (!browser.canUse('object-fit'))
				(function() {

					$('.banner .image, .spotlight .image').each(function() {

						var $this = $(this),
							$img = $this.children('img'),
							positionClass = $this.parent().attr('class').match(/image-position-([a-z]+)/);

						// Set image.
							$this
								.css('background-image', 'url("' + $img.attr('src') + '")')
								.css('background-repeat', 'no-repeat')
								.css('background-size', 'cover');

						// Set position.
							switch (positionClass.length > 1 ? positionClass[1] : '') {

								case 'left':
									$this.css('background-position', 'left');
									break;

								case 'right':
									$this.css('background-position', 'right');
									break;

								default:
								case 'center':
									$this.css('background-position', 'center');
									break;

							}

						// Hide original.
							$img.css('opacity', '0');

					});

				})();

	// Smooth scroll.
		$('.smooth-scroll').scrolly();
		$('.smooth-scroll-middle').scrolly({ anchor: 'middle' });

	// Wrapper.
		$wrapper.children()
			.scrollex({
				top:		'30vh',
				bottom:		'30vh',
				initialize:	function() {
					$(this).addClass('is-inactive');
				},
				terminate:	function() {
					$(this).removeClass('is-inactive');
				},
				enter:		function() {
					$(this).removeClass('is-inactive');
				},
				leave:		function() {

					var $this = $(this);

					if ($this.hasClass('onscroll-bidirectional'))
						$this.addClass('is-inactive');

				}
			});

	// Items.
		$('.items')
			.scrollex({
				top:		'30vh',
				bottom:		'30vh',
				delay:		50,
				initialize:	function() {
					$(this).addClass('is-inactive');
				},
				terminate:	function() {
					$(this).removeClass('is-inactive');
				},
				enter:		function() {
					$(this).removeClass('is-inactive');
				},
				leave:		function() {

					var $this = $(this);

					if ($this.hasClass('onscroll-bidirectional'))
						$this.addClass('is-inactive');

				}
			})
			.children()
				.wrapInner('<div class="inner"></div>');

	// Gallery.
		$('.gallery')
			.wrapInner('<div class="inner"></div>')
			.prepend(browser.mobile ? '' : '<div class="forward"></div><div class="backward"></div>')
			.scrollex({
				top:		'30vh',
				bottom:		'30vh',
				delay:		50,
				initialize:	function() {
					$(this).addClass('is-inactive');
				},
				terminate:	function() {
					$(this).removeClass('is-inactive');
				},
				enter:		function() {
					$(this).removeClass('is-inactive');
				},
				leave:		function() {

					var $this = $(this);

					if ($this.hasClass('onscroll-bidirectional'))
						$this.addClass('is-inactive');

				}
			})
			.children('.inner')
				//.css('overflow', 'hidden')
				.css('overflow-y', browser.mobile ? 'visible' : 'hidden')
				.css('overflow-x', browser.mobile ? 'scroll' : 'hidden')
				.scrollLeft(0);

		// Style #1.
			// ...

		// Style #2.
			$('.gallery')
				.on('wheel', '.inner', function(event) {

					var	$this = $(this),
						delta = (event.originalEvent.deltaX * 10);

					// Cap delta.
						if (delta > 0)
							delta = Math.min(25, delta);
						else if (delta < 0)
							delta = Math.max(-25, delta);

					// Scroll.
						$this.scrollLeft( $this.scrollLeft() + delta );

				})
				.on('mouseenter', '.forward, .backward', function(event) {

					var $this = $(this),
						$inner = $this.siblings('.inner'),
						direction = ($this.hasClass('forward') ? 1 : -1);

					// Clear move interval.
						clearInterval(this._gallery_moveIntervalId);

					// Start interval.
						this._gallery_moveIntervalId = setInterval(function() {
							$inner.scrollLeft( $inner.scrollLeft() + (5 * direction) );
						}, 10);

				})
				.on('mouseleave', '.forward, .backward', function(event) {

					// Clear move interval.
						clearInterval(this._gallery_moveIntervalId);

				});

		// Lightbox.
			$('.gallery.lightbox')
				.on('click', 'a', function(event) {

					var $a = $(this),
						$gallery = $a.parents('.gallery'),
						$modal = $gallery.children('.modal'),
						$modalImg = $modal.find('img'),
						href = $a.attr('href');

					// Not an image? Bail.
						if (!href.match(/\.(jpg|gif|png|mp4)$/))
							return;

					// Prevent default.
						event.preventDefault();
						event.stopPropagation();

					// Locked? Bail.
						if ($modal[0]._locked)
							return;

					// Lock.
						$modal[0]._locked = true;

					// Set src.
						$modalImg.attr('src', href);

					// Set visible.
						$modal.addClass('visible');

					// Focus.
						$modal.focus();

					// Delay.
						setTimeout(function() {

							// Unlock.
								$modal[0]._locked = false;

						}, 600);

				})
				.on('click', '.modal', function(event) {

					var $modal = $(this),
						$modalImg = $modal.find('img');

					// Locked? Bail.
						if ($modal[0]._locked)
							return;

					// Already hidden? Bail.
						if (!$modal.hasClass('visible'))
							return;

					// Lock.
						$modal[0]._locked = true;

					// Clear visible, loaded.
						$modal
							.removeClass('loaded')

					// Delay.
						setTimeout(function() {

							$modal
								.removeClass('visible')

							setTimeout(function() {

								// Clear src.
									$modalImg.attr('src', '');

								// Unlock.
									$modal[0]._locked = false;

								// Focus.
									$body.focus();

							}, 475);

						}, 125);

				})
				.on('keypress', '.modal', function(event) {

					var $modal = $(this);

					// Escape? Hide modal.
						if (event.keyCode == 27)
							$modal.trigger('click');

				})
				.prepend('<div class="modal" tabIndex="-1"><div class="inner"><img src="" /></div></div>')
					.find('img')
						.on('load', function(event) {

							var $modalImg = $(this),
								$modal = $modalImg.parents('.modal');

							setTimeout(function() {

								// No longer visible? Bail.
									if (!$modal.hasClass('visible'))
										return;

								// Set loaded.
									$modal.addClass('loaded');

							}, 275);

						});

})(jQuery);